### PR TITLE
[Transform] add method to collect deprecations from a transform configuration

### PR DESCRIPTION
**Pull request by hendrikmuhs**
add checkForDeprecations to TransformConfig and implement deprecation checks

Notes:
 - I refactored/moved some classes to re-use them in transform, see 1st commit: https://github.com/hendrikmuhs/elasticsearch/commit/d2938d94718c7f558948d28cc2eb18754ddef754: 
   - `DeprecationIssue` from the deprecations plugin /CC @jakelandis 
   - `LoggingDeprecationAccumulationHandler` from ML (semantic move inside x-pack core)
 - the real changes start at commit 2

Approach:

I am using a similar approach to ML, `checkForDeprecations` parses the configuration again, but instead of the ordinary deprecation logger re-uses the collector from ML datafeeds.

Todo:

 - integrate into deprecation API (followup PR?)

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/QueryConfig.java [hunk 1]
```diff
@@ -124,4 +142,37 @@ public ActionRequestValidationException validate(ActionRequestValidationExceptio
...
+                new DeprecationIssue(
+                    Level.CRITICAL,
+                    "Transform [" + id + "] uses deprecated query options",
+                    TransformDeprecations.QUERY_BREAKING_CHANGES_URL,
+                    deprecationMessage,
+                    false,
+                    null
+                )
+            );
+        });
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/QueryConfig.java:175]: ok, I like this way better than the way ML is doing it (these list of strings, etc.). Good refactoring I think++

**Review by benwtrent — APPROVED** [review]
I am fine with the integration into the deprecation API to be in a follow PR, especially since this PR has a larger refactor with moving things around. 

Would be good to jake or another es/data-managment team member to bless this PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/TransformDeprecations.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+package org.elasticsearch.xpack.core.transform;
+
+public class TransformDeprecations {
+    public static final String BREAKING_CHANGES_BASE_URL =
+        "https://www.elastic.co/guide/en/elasticsearch/reference/master/migrating-8.0.html";
+
+    public static final String QUERY_BREAKING_CHANGES_URL =
+        "https://www.elastic.co/guide/en/elasticsearch/reference/master/migrating-8.0.html#breaking_80_search_changes";
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/TransformDeprecations.java:15]: Should this be expressed as:
```
BREAKING_CHANGES_BASE_URL + "#breaking_80_search_changes"
```
?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/TransformDeprecations.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+    public static final String AGGS_BREAKING_CHANGES_URL =
+        "https://www.elastic.co/guide/en/elasticsearch/reference/master/migrating-8.0.html#breaking_80_aggregations_changes";
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/TransformDeprecations.java:18]: Should this be expressed as:
```
BREAKING_CHANGES_BASE_URL + "#breaking_80_aggregations_changes"
```
?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/QueryConfig.java [hunk 4]
```diff
@@ -98,6 +102,20 @@ public static QueryConfig fromXContent(final XContentParser parser, boolean leni
         return new QueryConfig(source, query);
     }
 
+    private static QueryBuilder queryFromXContent(
+        Map<String, Object> source,
+        NamedXContentRegistry namedXContentRegistry,
+        DeprecationHandler deprecationHandler
+    ) throws IOException {
+        QueryBuilder query = null;
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/QueryConfig.java:110]: Is this local variable needed at all?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/pivot/AggregationConfig.java [hunk 5]
```diff
@@ -131,6 +134,22 @@ public static AggregationConfig fromXContent(final XContentParser parser, boolea
         return new AggregationConfig(source, aggregations);
     }
 
+    private static AggregatorFactories.Builder aggregationsFromXContent(
+        Map<String, Object> source,
+        NamedXContentRegistry namedXContentRegistry,
+        DeprecationHandler deprecationHandler
+    ) throws IOException {
+        AggregatorFactories.Builder aggregations = null;
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/pivot/AggregationConfig.java:142]: Is this local variable needed?

**Review by przemekwitek — APPROVED** [review]
LGTM

Just 4 nits

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/QueryConfig.java (hunk 4 above)
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/QueryConfig.java:110]: yes, for lenient parsing we only log the exception and query remains `null`. I could also have 2 return's, but I think 1 exit is better than 2.

**Review by masseyke — APPROVED** [review]
Hendrick asked me to review just the part about moving DeprecationIssue to x-pack core. That makes sense to me.

**Pull request merged**